aws/rust-runtime/aws-auth/src/provider/cache.rs

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+use crate::provider::CredentialsResult;
+use crate::Credentials;
+use std::future::Future;
+use std::sync::Arc;
+use std::time::{Duration, SystemTime};
+use tokio::sync::{OnceCell, RwLock};
+
+#[derive(Clone)]
+pub(super) struct Cache {
+    margin: Duration,
+    value: Arc<RwLock<OnceCell<Credentials>>>,
+}
+
+impl Cache {
+    pub fn new(margin: Duration) -> Cache {
+        Cache {
+            margin,
+            value: Arc::new(RwLock::new(OnceCell::new())),
+        }
+    }
+
+    #[cfg(test)]
+    async fn get(&self) -> Option<Credentials> {
+        self.value.read().await.get().cloned()
+    }
+
+    pub async fn refresh<F, Fut>(&self, f: F) -> CredentialsResult
+    where
+        F: FnOnce() -> Fut,
+        Fut: Future<Output = CredentialsResult>,
+    {
+        let lock = self.value.read().await;
+        let future = lock.get_or_try_init(f);
+        future.await.map(|creds| creds.clone())
+    }
+
+    /// If the credentials are expired, clears the cache. Otherwise, yields the current credentials value.
+    pub async fn yield_or_clear_if_expired(&self, now: SystemTime) -> Option<Credentials> {
```

Comment:
I think we'll probably need a way to also explicitly expire credentials.

Reply:
What is the use-case for this?

Reply:
credentials can become invalidated (eg. manually invalidated session credentials)